src/util/result.h

```diff
@@ -6,37 +6,117 @@
 #define BITCOIN_UTIL_RESULT_H
 
 #include <util/translation.h>
-#include <variant>
+#include <optional>
 
-/*
- * 'BResult' is a generic class useful for wrapping a return object
- * (in case of success) or propagating the error cause.
-*/
-template<class T>
-class BResult {
-private:
-    std::variant<bilingual_str, T> m_variant;
+namespace util {
+
+//! Function result type intended for high-level functions that return error and
+//! warning strings in addition to normal result types.
+//!
+//! The Result<T> class is meant to be a drop-in replacement for
+//! std::optional<T> except it has additional methods to return error and
+//! warning strings for error reporting. Also, unlike std::optional, in order to
+//! support error handling in cases where callees need to pass additional
+//! information about failures back to callers, Result<T> objects can always be
+//! dereferenced regardless of the function's success or failure.
+//!
+//! This class is not intended to be used by low-level functions that do not
+//! return error or warning strings. These functions should use plain
+//! std::optional or std::variant types instead.
+//!
+//! See unit tests in result_tests.cpp for example usages.
+template<typename T>
+class Result;
+
+//! Tag types for result constructors.
+struct Error{};
+struct ErrorChain{};
+
+//! Result<void> specialization. Only returns errors and warnings, no values.
+template<>
+class Result<void>
+{
+protected:
+    std::vector<bilingual_str> m_errors;
```

Comment:
how about moving the members into a single `std::unique_ptr`, something like this:

```cpp
protected:
    struct ErrorsAndWarnings {
            std::vector<bilingual_str> m_errors;
            std::vector<bilingual_str> m_warnings;
    };
    std::unique_ptr<ErrorsAndWarnings> m_errors_and_warnings{};
```

This has the advantage that the default case when no error/warning happens is really fast: no temporary `std::vector` need to be constructed. Also, `sizeof()` is much smaller, only a single pointer. Then `Result` is also noncopyable. I think this would be an advantage, because usually these are supposed to be moved as the return value, and not copied.

Reply:
> how about moving the members into a single `std::unique_ptr`, something like this:

This is a great idea, and implemented in #25665